File: 00-code(源代码)/src/com/hi/dhl/algorithms/leetcode/_92/java/Solution.java
package com.hi.dhl.algorithms.leetcode._92.java;

import com.hi.dhl.algorithms.model.ListNode;

/**
 * <pre>
 *     author: dhl
 *     date  : 2022/6/30
 *     desc  :
 * </pre>
 */
class Solution {
    public ListNode reverseBetween(ListNode head, int left, int right) {
        ListNode dummyNode = new ListNode(0);
        dummyNode.next = head;
        ListNode p = dummyNode;
        int m = left;
        while (--m != 0) {
            p = p.next;
        }
        ListNode q = p.next;
        int k = right - left;
        while (k-- != 0) {
            ListNode node = q.next;
            q.next = q.next.next;
            node.next = p.next;
            p.next = node;
        }
        return dummyNode.next;
    }
}

File: 00-code(源代码)/src/com/hi/dhl/algorithms/leetcode/_92/kotlin/Solution.kt
package com.hi.dhl.algorithms.leetcode._92.kotlin

import com.hi.dhl.algorithms.model.ListNode

/**
 * <pre>
 *     author: dhl
 *     date  : 2022/6/30
 *     desc  :
 * </pre>
 */
class Solution {
    fun reverseBetween(head: ListNode?, left: Int, right: Int): ListNode? {
        val dummyNode = ListNode(0)
        dummyNode.next = head
        var p = dummyNode
        var m = left
        while (--m != 0) {
            p = p.next
        }
        var q = p.next
        var k = right - left
        while (k-- != 0) {
            val node = q.next
            q.next = q.next.next
            node.next = p.next
            p.next = node
        }
        return dummyNode.next
    }
}